lib/ansible/modules/cloud/amazon/rds_instance.py

```diff
@@ -0,0 +1,1236 @@
+    retry(waiter.wait)(WaiterConfig={'Delay': 20, 'MaxAttempts': 160}, DBInstanceIdentifier=db_instance_id)
+
+
+def wait_for_status(client, module, db_instance_id, waiter_name):
+    waiter_expected_status = {
+        'db_instance_deleted': 'deleted',
+        'db_instance_stopped': 'stopped',
+    }
+    expected_status = waiter_expected_status.get(waiter_name, 'available')
+
+    if expected_status == 'available':
+        extra_retry_codes = ['DBInstanceNotFound']
+    else:
+        extra_retry_codes = []
+
+    for attempt_to_wait in range(0, 10):
+        try:
+            wait(client, db_instance_id, waiter_name, extra_retry_codes)
+            break
+        except WaiterError as e:
+            # Instance may be renamed and AWSRetry doesn't handle WaiterError
+            if e.last_response.get('Error', {}).get('Code') == 'DBInstanceNotFound':
+                sleep(10)
+                continue
+            module.fail_json_aws(e, msg="Error while waiting for DB instance {0} to be {1}".format(db_instance_id, expected_status))
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Unexpected error while waiting for DB instance {0} to be {1}".format(
+                db_instance_id, expected_status)
+            )
+
+
+def ensure_tags(client, module, resource_arn, existing_tags, tags, purge_tags):
+    if tags is None:
+        return False
+
+    tags_to_add, tags_to_remove = compare_aws_tags(existing_tags, tags, purge_tags)
+    changed = bool(tags_to_add or tags_to_remove)
+
+    if tags_to_add:
+        call_method(
+            client, module, method_name='add_tags_to_resource',
+            parameters={'ResourceName': resource_arn, 'Tags': ansible_dict_to_boto3_tag_list(tags_to_add)}
+        )
+    if tags_to_remove:
+        call_method(
+            client, module, method_name='remove_tags_from_resource',
+            parameters={'ResourceName': resource_arn, 'TagKeys': tags_to_remove}
+        )
+
+    return changed
+
+
+def update_instance(client, module, instance, instance_id):
+    changed = False
+
+    # Get newly created DB instance
+    if not instance:
+        instance = get_instance(client, module, instance_id)
+
+    # Check tagging/promoting/rebooting/starting/stopping instance
+    changed |= ensure_tags(
+        client, module, instance['DBInstanceArn'], instance['Tags'], module.params['tags'], module.params['purge_tags']
+    )
+    changed |= promote_replication_instance(client, module, instance, module.params['promote'])
+    changed |= update_instance_state(client, module, instance, module.params['state'])
+
+    return changed
+
+
+def promote_replication_instance(client, module, instance, promote):
+    if instance.get('ReplicationSourceIdentifier') and promote:
+        call_method(
+            client, module, method_name='promote_read_replica_db_instance',
+            parameters={'DBInstanceIdentifier': instance['DBInstanceIdentifier']}
+        )
```

Comment:
It looks like if an instance is promoted, call_method will attempt to wait if the use sets wait=True, but I don't see a waiter in the RDS docs or in the custom waiters that would cover that. Is that true?

Reply:
I tested this out (and will add it to the test suite). The db_instance_available built-in waiter is used and works as expected https://boto3.readthedocs.io/en/latest/reference/services/rds.html#RDS.Waiter.DBInstanceAvailable.